### PR TITLE
SAMZA-1812: Added configuration for changelog to local store backed tables

### PR DESCRIPTION
**Pull request by weisong44**
Currently changelog for tables can be supported implicitly by adding user defined configuration, to be more user friendly we want to expose it though table API. This is applicable to RocksDB and in memory table implementation.
 - By default, changelog is enabled with auto-generated name <job-name><job-id>-table<table-id>
 - Added ability to disable changelog
 - Added ability to set changelog stream name, by default
 - Added validation of changelog stream name, and modify to confirm to Kafka topic name spec
 - Added configuration for replication factor
 - Disable changelog when user enables side input

### CODE REVIEW

File: samza-api/src/main/java/org/apache/samza/table/TableProvider.java [hunk 1]
```diff
@@ -31,6 +32,12 @@
  */
 @InterfaceStability.Unstable
 public interface TableProvider {
+  /**
+   * Initialize TableProvider with job configuration
```
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-api/src/main/java/org/apache/samza/table/TableProvider.java]: Config is contained in ContainerContext, but the problem is when we are generating config, the other method was not called. And we couldn't get hold of the context objects too.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java [hunk 2]
```diff
@@ -36,11 +36,20 @@
  */
 abstract public class BaseLocalStoreBackedTableDescriptor<K, V, D extends BaseLocalStoreBackedTableDescriptor<K, V, D>>
     extends BaseTableDescriptor<K, V, D> {
+
+  static final public String INTERNAL_ENABLE_CHANGELOG = "internal.enable.changelog";
```
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: The reason is that when we create the TableSpec, we don't know the job name and id, so the info has to be stashed for later use during planning phase.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java [hunk 3]
```diff
@@ -72,6 +131,14 @@ protected void validate() {
           String.format("Invalid side input configuration for table: %s. " +
               "Both side inputs and the processor must be provided", tableId));
     }
+    if (!enableChangelog) {
+      Preconditions.checkState(changelogStream == null,
```
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Here they actually need to be null

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java [hunk 4]
```diff
@@ -47,26 +46,20 @@
  * the table provider will not manage the lifecycle of the backing
  * stores.
  */
-abstract public class BaseLocalStoreBackedTableProvider implements TableProvider {
-
-  protected final Logger logger = LoggerFactory.getLogger(getClass());
+abstract public class BaseLocalStoreBackedTableProvider extends BaseTableProvider {
```
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java]: +1

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java [hunk 5]
```diff
@@ -47,26 +46,20 @@
...
 
-  protected final TableSpec tableSpec;
+  final static private Pattern INVALID_CHANGELOG_CHARS = Pattern.compile("[^a-zA-Z0-9_-]+");
```
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java]: +1

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java [hunk 6]
```diff
@@ -36,11 +36,20 @@
...
 abstract public class BaseLocalStoreBackedTableDescriptor<K, V, D extends BaseLocalStoreBackedTableDescriptor<K, V, D>>
     extends BaseTableDescriptor<K, V, D> {
+
+  static final public String INTERNAL_ENABLE_CHANGELOG = "internal.enable.changelog";
+  static final public String INTERNAL_CHANGELOG_STREAM = "internal.changelog.stream";
+  static final public String INTERNAL_CHANGELOG_REPLICATION_FACTOR = "internal.changelog.replication.factor";
+
   protected List<String> sideInputs;
   protected SideInputsProcessor sideInputsProcessor;
+  protected boolean enableChangelog = true;
```
Comment (bharathkk) [inline samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Are we taking a conscious decision of having changelog enabled by default for tables? The current behavior of user-defined stores in high-level or low level has changelog disabled by default.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java [hunk 7]
```diff
@@ -49,6 +58,8 @@ public BaseLocalStoreBackedTableDescriptor(String tableId) {
 
   public D withSideInputs(List<String> sideInputs) {
     this.sideInputs = sideInputs;
+    this.enableChangelog = false;
+    this.changelogStream = null;
```
Comment (bharathkk) [inline samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: we need to set the `changelogReplicationFactor` to `null` as well.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java [hunk 8]
```diff
@@ -57,9 +68,57 @@ public D withSideInputsProcessor(SideInputsProcessor sideInputsProcessor) {
...
+   */
+  public D withChangelogStream(String changelogStream) {
+    this.changelogStream = changelogStream;
+    return (D) this;
+  }
+
+  /**
+   * Refer to <code>stores.store-name.changelog.replication.factor</code> in Samza configuration guide
+   *
+   * @param replicationFactor replilcation factor
```
Comment (bharathkk) [inline samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: nit: s/replilcation/replication

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bharathkk) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: +1.
Maybe only check for `changelogStream` to be non-null in the case of `enableChangelog` is `true`. `changelogReplicationFactor` is an optional configuration.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bharathkk) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: all the more reason to use the descriptor in the planning phase so that we can get rid of the internal placeholders for passing configurations around. Looking forward to the refactor of using descriptors in the planning phase :)

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: In fact both are optional, the checks were added just to avoid any errors from user's side.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: :)

File: samza-api/src/main/java/org/apache/samza/table/TableProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-api/src/main/java/org/apache/samza/table/TableProvider.java]: +1 will make the changes

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 6 above)
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Well, this is a discrepancy already. But I think it is probably more advantageous to keep it enabled by default. We have an option to disable it and documented in the javadoc. What do you think?

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bharathkk) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Then, we are not truly enabling changelog by default right? We still require applications to set the `changelogStream`. Is that correct?

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 6 above)
Comment (bharathkk) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Sure, I don't have a strong opinion on that. However, what does it mean to have a changelog for in-memory stores or RocksDb that is in-memory? 
Also, refer to my comments below on both `changelogStream` and `changelogReplicationFactor` both being optional parameters.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java [hunk 9]
```diff
@@ -114,6 +110,30 @@ public Table getTable() {
           SerdeUtils.serialize("Side Inputs Processor", tableSpec.getSideInputsProcessor()));
     }
 
+    // Changelog configuration
+    String enableChangelog = tableSpec.getConfig().get(BaseLocalStoreBackedTableDescriptor.INTERNAL_ENABLE_CHANGELOG);
+    if ("true".equalsIgnoreCase(enableChangelog)) {
```
Comment (bharathkk) [inline samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java]: Not sure if this is out of scope for this PR. But, I guess since `TableSpec` is immutable, you can have the config inside TableSpec as a `MapConfig` instead of a `Map<String, String>` so that you can get the type information for free :)

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java [hunk 10]
```diff
@@ -57,9 +70,57 @@ public D withSideInputsProcessor(SideInputsProcessor sideInputsProcessor) {
...
+   * generated in format [job-name]-[job-id]-table-[table-id]. For both auto-generated
+   * and user specified changelog stream, invalid characters will be automatically
+   * replaced with '-' to confirm with pattern { @literal [a-zA-Z0-9_-]+ }
+   *
+   * Refer to <code>stores.store-name.changelog</code> in Samza configuration guide
+   *
+   * @param changelogStream changelog stream name
+   * @return this table descriptor instance
+   */
+  public D withChangelogStream(String changelogStream) {
```
Comment (bharathkk) [inline samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Do we have a use-case where it would be useful to override the changelogStream? 
Also, can we not rewrite the changelog stream names here but throw when the names don't conform to the pattern `@literal [a-zA-Z0-9_-]+`

Comment (bharathkk) [inline samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Do we have a use-case for overriding the auto-generated `changelogStream`? 
Is it possible to throw an exception instead of rewriting the changelog stream if it doesn't conform with the pattern `[a-zA-Z0-9_-]+`?

Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Since this is more of an internal feature, wouldn't it more user friendly to auto-repair it?

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: No, the check is in effect only if user DISABLES changelog. Primarily to avoid user oversight. When enabled, both parameters are optional.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java (hunk 9 above)
Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableProvider.java]: Yeah, I'd prefer to take this to another PR.

File: samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java (hunk 10 above)
Comment (prateekm) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: @weisong44 I'd prefer to throw an exception here as well. If the auto-generated name is invalid, it would imply that the job name is invalid. Job name is used in other places to create streams (e.g., coordinator, checkpoint) so that'd still be an issue, and users should fix that. 

In general, I'd prefer to not automatically rewrite names - it causes confusion since the name you'd expect from code or convention doesn't match the actual name. For example, this is an issue with some metrics reporters that "normalize" dashes in the metric name to underscores - it's very confusing to figure out the actual metric name.

Comment (weisong44) [reply samza-kv/src/main/java/org/apache/samza/storage/kv/BaseLocalStoreBackedTableDescriptor.java]: Based on what we discussed

File: samza-core/src/main/java/org/apache/samza/operators/BaseTableDescriptor.java [hunk 11]
```diff
@@ -45,7 +45,7 @@
 
   /**
    * Constructs a table descriptor instance
-   * @param tableId Id of the table
+   * @param tableId Id of the table, it must confirm to pattern { @literal [\d\w-_]+ }
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/BaseTableDescriptor.java]: Maybe `\\d\\w` to match java regex conventions?

File: samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java [hunk 12]
```diff
@@ -145,12 +146,11 @@ public void setDefaultSerde(Serde<?> serde) {
 
   @Override
   public <K, V> Table<KV<K, V>> getTable(TableDescriptor<K, V, ?> tableDesc) {
+    Preconditions.checkState(isValidTableId(tableDesc.getTableId()), String.format(
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java]: Maybe validate changelog and side input names using the Stream ID pattern too?
